### PR TITLE
Rayleigh cdf always nonnegative

### PR DESCRIPTION
**Pull request by aplavin**
fix https://github.com/JuliaStats/Distributions.jl/issues/1371

### CODE REVIEW

File: src/univariate/continuous/rayleigh.jl [hunk 1]
```diff
@@ -83,7 +83,7 @@ function logccdf(d::Rayleigh, x::Real)
 end
 ccdf(d::Rayleigh, x::Real) = exp(logccdf(d, x))
 
-cdf(d::Rayleigh, x::Real) = -expm1(logccdf(d, x))
+cdf(d::Rayleigh, x::Real) = false - expm1(logccdf(d, x))  # zero (false) ensures that cdf cannot be -0
```
Comment (devmotion) [inline src/univariate/continuous/rayleigh.jl:86]: I would prefer something that fixes the issue more directly - this seems like a quite weird workaround. The underlying problem here is that the implementation of `expm1` in base returns
```julia
julia> expm1(0.0)
0.0

julia> expm1(-0.0)
-0.0
```
Since
```julia
julia> iszero(-0.0)
true

julia> iszero(0.0)
true
```
both positive and negative zero are proper zeros, I'm not completely sure why a negative zero would be a problem here. Do you have a MWE where it causes problems?

If we want to change it maybe we could just use
```suggestion
# we ensure that the cdf cannot be -0
cdf(d::Rayleigh, x::Real) = max(- expm1(logccdf(d, x)), 0)
```

Comment (aplavin) [reply src/univariate/continuous/rayleigh.jl:86]: See the linked issue for motivation. A simple example:
```
julia> using Distributions
julia> d = Rayleigh(1)
# nonzero works fine:
julia> rand(Binomial(100, cdf(d, 0.1)))
2
# zero doesn't:
julia> rand(Binomial(100, cdf(d, 0)))
ERROR: InexactError: trunc(Int64, -Inf)
<...>
# but it would, if cdf didn't return negative zero:
julia> rand(Binomial(100, 0.0))
0
```

File: test/continuous.jl [hunk 2]
```diff
@@ -98,3 +98,7 @@ end
     @test var(d) ≈ δ * α^2 / g^3
     @test skewness(d) ≈ 3β/(α*sqrt(δ*g))
 end
+
+@testset "edge cases" begin
+    @test cdf(Rayleigh(1), 0) === 0.0  # not -0
```
Comment (devmotion) [inline test/continuous.jl:103]: Maybe it's helpful to mention the issue here?
```suggestion
    # issue #1371: cdf should not return -0.0
    @test cdf(Rayleigh(1), 0) === 0.0
```

File: src/univariate/continuous/rayleigh.jl (hunk 1 above)
Comment (aplavin) [reply src/univariate/continuous/rayleigh.jl]: Actually, a simpler solution is possible. Changed.

File: src/univariate/continuous/rayleigh.jl [hunk 3]
```diff
@@ -79,7 +79,7 @@ end
 
 function logccdf(d::Rayleigh, x::Real)
     z = - x^2 / (2 * d.σ^2)
-    return x > 0 ? z : isnan(x) ? oftype(z, NaN) : zero(x)
+    return x >= 0 ? z : isnan(x) ? oftype(z, NaN) : zero(x)
```
Comment (devmotion) [inline src/univariate/continuous/rayleigh.jl:82]: Oh, actually we should use
```suggestion
    return x >= 0 ? z : isnan(x) ? oftype(z, NaN) : zero(z)
```
here to ensure type stability. It shouldn't matter though if we use `>` or `>=`, should it?

Comment (aplavin) [reply src/univariate/continuous/rayleigh.jl]: Almost - it should be negative zero(z), see updated commit.

File: src/univariate/continuous/rayleigh.jl [hunk 4]
```diff
@@ -79,7 +79,7 @@ end
 
 function logccdf(d::Rayleigh, x::Real)
     z = - x^2 / (2 * d.σ^2)
-    return x > 0 ? z : isnan(x) ? oftype(z, NaN) : zero(x)
+    return x >= 0 ? z : isnan(x) ? oftype(z, NaN) : -zero(z)
```
Comment (devmotion) [inline src/univariate/continuous/rayleigh.jl:82]: Yeah, this is simpler (thought about it actually but assumed that it doesn't matter if we return negative or positive zeros for `logccdf`). Could you add the comment about positive cdf here? Otherwise I assume it can be a bit unclear for a reader why `-zero(z)` is used here.

**Comment by devmotion** [comment]
It seems one of the tests fails? https://github.com/JuliaStats/Distributions.jl/runs/4279041031?check_suite_focus=true#step:6:469

**Comment by aplavin** [comment]
fixed

**Review by devmotion — APPROVED** [review]
Thanks, looks good to me! The test error is unrelated (and there's an open PR that fixes it).
